[Creation] Add creation capability

VWoeltjen opened the pull request:
Add a `creation` capability which can be used to instantiate new domain objects using an existing domain object. Created objects are not persisted nor added to any objects composition, allowing these steps to be handled separately (using the created object's `persistence` and the parent object's `composition` capabilities, respectively), per #215.

Summary of changes:
* Add `creation` capability, with a `create` method that generates a new DomainObject instance from the provided model.
* Move `uuid` library to core, from commonUI
* Update `creationService` to use this capability, instead of interacting with `persistenceService` directly.

@akhenry  - as discussed, there's some refactoring to duplicate as implemented in #249 that could use this capability. Hadn't expected to address this so soon, but after evaluating a few strategies believe this will be useful for #245. I see three ways to reconcile this with #249 - I'm fine with any option, let me know which you prefer.

1. Merge #249 first, then I update this PR to reflect the desired refactoring
2. Merge this PR first, update #249 to use it
3. Handle both separately, file a follow up issue to reconcile (believe this change set will not break changes from #249, so this should be an option.)

All else being equal, I'd go with Option 3 to maintain forward progress.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Expect to pass code review? Y

akhenry commented:
Inclined to go with option 3 for now and defer to Banks. I don't think I will get the refactoring necessary for this into Ballard.

I'd like to think about the CopyService a little more as well. Maybe it would be desirable to separate duplication and persistence of objects in the same way they have now been separated for creation. I don't have a specific use case in mind, but I wonder if it might prove useful in future, potentially in edit mode. Also, it would simplify the CopyService

akhenry commented on platform/core/src/capabilities/CreationCapability.js:
```diff
@@ -0,0 +1,79 @@
+
+        return CreationCapability;
+    }
+);
```
It feels a little funny implementing this as a capability. My understanding is that a capability is associated with an instance of a domain object, and is retrieved through getCapability or useCapability on that domain object instance. It doesn't feel right that creating a new domain object is done from an existing domain object instance, but then there is no relationship enforced between the original and new domain objects? It might make sense doing it this way if this were a "createChild" capability that automatically composed the new object as well as creating it, for example.

My natural assumption was that creation would be done through the existing CreationService.createObject method, but that the persistence step would be removed from there, so that creation and persistence were treated as separate operations. CreationService.createObject would become a factory method, and then new objects returned would be persisted via newObject.useCapability("persistence"). This approach would require refactoring of existing code though.

VWoeltjen replied:
I agree that this is a little funny; the rationale for doing it this way is to support multiple persistence spaces, #245. I've opened a PR for this early (#257) which has some additional changes here.

To frame this a little, some of the practical use cases for multiple persistence spaces are:

* Providing domain object hierarchies for multiple missions, where each mission's hierarchy is stored at a different URL (to allow for permissions management.)
* Separating private ("My Items") from public information ("Mission Items" or similar); again, may be stored at different URLs for simplified permissions management
* (Less immediate, may not take this approach) Providing read/write access for "domain objects" which reside in a different application (e.g. xGDS traverse plans.)

In all cases, the desired behavior is that the back-end space that you persist to depends on the location of the created object.

A case where this gets complicated is Duplicate. To avoid leaving partially-duplicated trees behind, we're persisting objects deeper in the tree first, ending with an persist of the destination of the Duplicate. As such, the location hierarchy isn't complete until the last object gets created; so, we need a short-hand to say "create this new object such that it will be persisted in the same space as this other object."

In other words, the specific details of creating a new domain object need to vary depending on where it will be created, even if it won't immediately be a child of that destination. Putting something in a capability is a handy way to allow behavior to vary per domain object. (Another constraint is introduced by the implementation approach for #245; when reading objects, you need to know which space to read from, and the easiest way to do that is to encode it in the identifier. To make this work, you need to know the space at instantiation time, because that is also where you'd generate the identifier. Sticking this in a capability makes it convenient to hide those details.)

We could also handle this in `creationService` (by changing semantics of `creationService.createObject(model, parent)` to mean the same thing as `parent.useCapability('creation', model)` would in this PR.) I tend to prefer capabilities for behavior that may vary per-object, because if these rules change (or if a different deployment wants to introduce different rules for some objects) this can be reasonably well-hidden by overriding or wrapping capabilities for objects of certain types (can also be done at the service layer via decoration, but you end up with a lot of iffing to recognize the cases you want to override, and it's just cleaner to handle that polymorphically)

Anyway, open to other options, just wanted to give some context for that decision

akhenry replied:
Hrm, ok I understand the rationale now, thanks. 
Here's a specific case where I think it gets tricky though -
* In new edit mode, the user can drag a plot onto another plot which has the effect of creating a new 'virtual' panel in memory. 
* The view switches to edit mode immediately, and the location of the new virtual panel is not determined until the user hits save, at which point the create dialog is launched allowing them to select a location in the object tree to save their new panel. 
* The two original plots might have been persisted in a remote elastic search instance, but the user might want to persist their new combined panel into My Items. 
* The virtual panel could be created initially using the persistence space of the base plot, but this would then need to be updated prior to persistence.

akhenry replied:
In short, it would be nice if it were possible to leave resolution of persistence space until its time to persist the object

VWoeltjen replied:
Thanks for raising that case - I hadn't considered this.

There's a similar problem for Move between spaces (see comments on #245); a lot of identifiers need to get rewritten for the Move to work. Think this can be solved as a Duplicate + Remove; I guess a Save As could be solved similarly.

Still, makes sense that persistence space should not need to be decided until you persist. This is at odds with encoding persistence spaces in identifiers, however (in the process of editing, you might be using this identifier in the composition of all sorts of other documents, so needing to rewrite that when you pick a persistence space becomes a pain.) Reassigning this back over to me to rethink this.

akhenry commented:
Comments inline. 

Can I just say how happy I am to have this functionality :) It's going to be useful for me in a number of places.

VWoeltjen commented:
Addressed comments from code review; summary of changes:

* Added an `instantiate` service that serves as a factory for domain object instances, allowing instantiation outside of the context of any specific domain object
* Left the capability for this in place (for cases where you do want to instantiate in the context of a specific domain object), but renamed to `instantiation` for consistency.
  * Think the nomenclature change is also slightly helpful in disambiguating - "creation" involves persistence, "instantiation" doesn't.
* Updated DomainObjectProvider to use `instantiate` for consistency (`instantiate` nicely pulls out the commonality between DomainObjectProvider and InstantiationCapability to facilitate this)

akhenry commented:
This looks great. 

I'm quite not clear how a newly instantiated object acquires a persistence space though. Will that happen automatically at persistence time based on the persistence space of its locating object, or does it happen at composition time, when the new objects gets added to an existing persisted object? If so, what happens if the new object is part of a chain of new objects that have not been persisted yet, and are currently unhooked from any real objects? (Thinking of bottom-up copy).

VWoeltjen commented:
Working that out in the context of #245; just want to introduce appropriate locations in the API to do so here.

For cases like Duplicate, the space would be determined within the `instantiation` capability of the destination object.

For cases like new edit mode, not sure yet. Things created via the `instantiate` service would be in the default persistence space (which, if you only have one persistence space, means you just don't have to worry about it); once saved to a new location, the tree would need to be re-written to use the space from the new location (_if_ that new location is in a different space). Think this can reuse code for however we solve Move for the analogous case (probably by implementing it as a Duplicate + Delete.) In that case, we'd want to make sure that the "delete" was lightweight (since those objects had never been persisted, it should be a no-op.)

akhenry commented on platform/core/test/capabilities/InstantiationCapabilitySpec.js:
```diff
@@ -0,0 +1,70 @@
+/*global define,Promise,describe,it,xdescribe,expect,beforeEach,waitsFor,jasmine*/
+
+/**
+ * ContextCapability. Created by vwoeltje on 11/6/14.
```
Copy-paste error?

akhenry commented:
Looks good to me.

## Reviewer Checklist

Changes appear to address issue? __Y__
Appropriate unit tests included? __Y__
Code style and in-line documentation are appropriate? __Y__
Commit messages meet standards? __Y__
